cloudnetwork/v1/types.go

```diff
@@ -0,0 +1,102 @@
+// the cluster network.
+// Note: this CRD is managed using a singleton cluster-wide CR, which is
+// called: `cluster`. This is enforced by the CRD's naming convention in
+// combination with the fact that the CRD is cluster-scoped.
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:openapi-gen=true
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:path=cloudprivateipconfig,scope=Cluster
+type CloudPrivateIPConfig struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// Specification of the desired private IP request(s).
+	// +kubebuilder:validation:Required
+	// +required
+	Spec CloudPrivateIPConfigSpec `json:"spec" protobuf:"bytes,2,opt,name=spec,casttype=CloudPrivateIPConfigSpec"`
+	// Observed status of CloudPrivateIPConfig. Read-only.
+	// +kubebuilder:validation:Optional
+	// +optional
+	Status CloudPrivateIPConfigStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status,casttype=CloudPrivateIPConfigStatus"`
+}
+
+// CloudPrivateIPConfigSpec contains the requested list of private IPs and their corresponding node assignment.
+// +k8s:openapi-gen=true
+type CloudPrivateIPConfigSpec struct {
+	// +kubebuilder:validation:Required
+	// +required
+	Items []CloudPrivateIPConfigSpecItem `json:"items" protobuf:"bytes,1,opt,name=items"`
+}
+
+// CloudPrivateIPConfigStatus contains the assigned list of private IPs and their corresponding node assignment.
+// +k8s:openapi-gen=true
+type CloudPrivateIPConfigStatus struct {
+	// +kubebuilder:validation:Required
+	// +required
+	Items []CloudPrivateIPConfigStatusItem `json:"items" protobuf:"bytes,1,opt,name=items"`
+}
+
+// CloudPrivateIPConfigSpecItem is a tuple consisting of a node name and IP address.
+// +k8s:openapi-gen=true
+type CloudPrivateIPConfigSpecItem struct {
+	// node specified the node name, as specified by the Kubernetes field: `node.metadata.name`
+	// +kubebuilder:validation:Required
+	// +required
+	Node string `json:"node" protobuf:"bytes,1,opt,name=node"`
+	// ip specified the IP address, can be IPv4 or IPv6
+	// +kubebuilder:validation:Required
+	// +required
+	IP string `json:"ip" protobuf:"bytes,2,opt,name=ip"`
+}
+
+// Status specifies the current status of the CloudPrivateIPConfigStatusItem
+type Status string
+
+const (
+	// Error defines a failed cloud assignment of the IP address requested.
+	Error Status = "ERROR"
+
+	// Pending defines a pending cloud assignment of the IP address requested.
+	Pending Status = "PENDING"
+
+	// Success defines a successful cloud assignment of the IP address requested.
+	Success Status = "SUCCESS"
+)
+
+// CloudPrivateIPConfigStatusItem contains the requested list of private IPs and their corresponding node assignment
+// together with each item's status.
+// +k8s:openapi-gen=true
+type CloudPrivateIPConfigStatusItem struct {
```

Comment:
The convention is for status to be Conditions, e.g. https://github.com/kubernetes/api/blob/master/core/v1/types.go#L2541

Reply:
ACK, will change to Conditions.